libraries/ESP8266WiFi/src/ESP8266WiFiSTA.h

```diff
@@ -93,13 +95,11 @@ class ESP8266WiFiSTAClass {
 
     public:
 
-#ifdef NO_EXTRA_4K_HEAP
-        bool beginWPSConfig(void);
-#else
-        inline bool beginWPSConfig(void) __attribute__((always_inline)) {
-            return WPS_is_unavailable_in_this_configuration__Please_check_FAQ_or_board_generator_tool();
+        inline bool beginWPSConfig(void) __attribute__((always_inline))
+        {
+            disable_extra4k_at_link_time(); // this call must always be inlined

```

Comment:
as far as i can tell this can never be inlined, unless we enable LTO?

Reply:
I have updated the code/comment (yet to push) as follow:
```
    protected:
    
        static void WPSStatusCB(wps_cb_status status);
        bool beginRealWPSConfig(void);

    public:

        inline bool beginWPSConfig(void) __attribute__((always_inline))
        {
            /*
             Some notes:

             This call to disable_extra4k_at_link_time() must not be
             "compiled-in" when WPS is not in use.  Calling it from
             beginRealWPSConfig() would always link it.

             'inline ...  attribute(always_inline)' is not required since gcc
             will optimize this function, but it is left as a reminder.

             Moving beginRealWPSConfig from ESP8266WiFiSTA.cpp to
             ESP8266WiFiSTA-WPS.cpp only allows to same some flash space.
            */
            disable_extra4k_at_link_time();

            return beginRealWPSConfig();
        }
```
When it is not inlined, or not optimized by gcc, or if `disable_extra4k_at_link_time()` is called from `beginRealWPSConfig()`, heap stack is always linked in.

Reply:
Okay, i finally understood why you were mentioning inlining in other comments and why you needed to make it inline. This is because Arduino IDE does not combine object files from libraries into static archives, it links them directly. This is in contrast with the core object files, which are first placed into an archive.

However, recent Arduino versions support putting library object files into .a archive before linking, which should remove the need for inlining. Please check https://github.com/arduino/Arduino/wiki/Arduino-IDE-1.5:-Library-specification and add `dot_a_linkage=true` to ESP8266WiFi library.properties file.

Reply:
It works with `dot_a_linkage=true` and the need for an `inline` function/method in the class is now useless.

Things are now more clear, thanks for this tip. I was not understanding in the first place why I had to use the inline trick and was wrongly thinking that objects files were stored in a library. I admit I don't often try to read the long-like-hell IDE commands.

But I had to cheat and create an empty archive `{tmp}/libraries/ESP8266WiFi/ESP8266WiFi.a` at the linker command request, while the build process put everything in `{tmp}/arduino.ar`.
It seems to be the same bug as [this esp32 one](https://github.com/espressif/arduino-esp32/issues/1049). This issue is still present in both latest nightly and latest beta.

Reply:
*edit2*: found it, it's `{archive_file_path}`. Now `arduino.ar` no longer exists, digging this out :)